File: src/lib/videoRuntime/layerTracks.ts
export type FrameIndex = number;

export interface LayerRange {
  startFrame: FrameIndex;
  endFrame: FrameIndex;
  /** If set, overrides payload.enabled for frames inside the range. */
  enabled?: boolean;
  /** If set, overrides payload.opacity (0..1) for frames inside the range. */
  opacity01?: number;
  /** If set, overrides payload.intensity for frames inside the range. */
  intensity?: number;
  /** Source media in-point (frame index). When set together with sourceOutFrame,
   *  the resolved source frame is clamped to [sourceInFrame, sourceOutFrame]. */
  sourceInFrame?: FrameIndex;
  /** Source media out-point (frame index). When set together with sourceInFrame,
   *  the resolved source frame is clamped to [sourceInFrame, sourceOutFrame]. */
  sourceOutFrame?: FrameIndex;
}

export interface LayerKeyframe {
  frame: FrameIndex;
  opacity01?: number;
  intensity?: number;
}

export interface LayerTrack {
  layerId: string;
  /** If true, layer is disabled outside any defined range. Default: true. */
  disableOutsideRanges?: boolean;
  ranges: LayerRange[];
  keyframes: LayerKeyframe[];
}

export function cloneLayerRange(range: LayerRange): LayerRange {
  return { ...range };
}

export function cloneLayerKeyframe(keyframe: LayerKeyframe): LayerKeyframe {
  return { ...keyframe };
}

export function cloneLayerTrack(track: LayerTrack): LayerTrack {
  return {
    ...track,
    ranges: track.ranges.map(cloneLayerRange),
    keyframes: track.keyframes.map(cloneLayerKeyframe),
  };
}

export function cloneLayerTracks(tracks: LayerTrack[]): LayerTrack[] {
  return tracks.map(cloneLayerTrack);
}

function clamp01(value: number): number {
  return Math.max(0, Math.min(1, value));
}

function findActiveRange(ranges: LayerRange[], frame: number): LayerRange | null {
  for (const r of ranges) {
    if (frame >= r.startFrame && frame <= r.endFrame) return r;
  }
  return null;
}

function findKeyframeBounds(keyframes: LayerKeyframe[], frame: number): { prev: LayerKeyframe | null; next: LayerKeyframe | null } {
  let prev: LayerKeyframe | null = null;
  let next: LayerKeyframe | null = null;
  for (const k of keyframes) {
    if (k.frame <= frame && (!prev || k.frame > prev.frame)) prev = k;
    if (k.frame >= frame && (!next || k.frame < next.frame)) next = k;
  }
  return { prev, next };
}

function lerpNumber(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

function interpolateKeyframes(keyframes: LayerKeyframe[], frame: number): { opacity01?: number; intensity?: number } {
  if (!keyframes.length) return {};
  const { prev, next } = findKeyframeBounds(keyframes, frame);
  if (!prev && !next) return {};
  if (prev && (!next || next.frame === prev.frame)) {
    return { opacity01: prev.opacity01, intensity: prev.intensity };
  }
  if (!prev && next) {
    return { opacity01: next.opacity01, intensity: next.intensity };
  }
  if (!prev || !next) return {};
  const span = Math.max(1, next.frame - prev.frame);
  const t = (frame - prev.frame) / span;
  const out: { opacity01?: number; intensity?: number } = {};
  if (typeof prev.opacity01 === "number" && typeof next.opacity01 === "number") {
    out.opacity01 = clamp01(lerpNumber(prev.opacity01, next.opacity01, t));
  }
  if (typeof prev.intensity === "number" && typeof next.intensity === "number") {
    out.intensity = lerpNumber(prev.intensity, next.intensity, t);
  }
  return out;
}

/**
 * Apply time-based tracks to a backend layer payload array.
 * This operates at the payload level to avoid breaking the canonical Layer->DTO mapping.
 */
export function applyLayerTracksToPayload(
  payload: Array<Record<string, unknown>>,
  tracks: LayerTrack[],
  frameIndex: number,
): Array<Record<string, unknown>> {
  if (!tracks.length) return payload;
  const trackByLayerId = new Map(tracks.map((t) => [t.layerId, t]));

  return payload.map((layer) => {
    const id = String(layer.id ?? "");
    const track = trackByLayerId.get(id);
    if (!track) return layer;

    const activeRange = findActiveRange(track.ranges, frameIndex);
    const disableOutside = track.disableOutsideRanges ?? true;

    const next: Record<string, unknown> = { ...layer };

    if (!activeRange && disableOutside) {
      next.enabled = false;
      return next;
    }

    if (activeRange) {
      if (typeof activeRange.enabled === "boolean") next.enabled = activeRange.enabled;
      if (typeof activeRange.opacity01 === "number") next.opacity = clamp01(activeRange.opacity01);
      if (typeof activeRange.intensity === "number") next.intensity = activeRange.intensity;
    }

    const kf = interpolateKeyframes(track.keyframes, frameIndex);
    if (typeof kf.opacity01 === "number") next.opacity = clamp01(kf.opacity01);
    if (typeof kf.intensity === "number") next.intensity = kf.intensity;

    return next;
  });
}

File: src-tauri/src/video_runtime/layer_tracks.rs
use crate::image_engine::EffectLayer;
use super::types::{LayerTrack, LayerRange, LayerKeyframe};

/// Maps a timeline frame index to the corresponding source (file) frame index
/// for a given layer range.
///
/// When both `source_in_frame` and `source_out_frame` are set on the range,
/// the formula is: `min(source_in + (timeline_frame - start_frame), source_out)`.
/// This clamps the source frame so it never exceeds `source_out_frame`.
///
/// When either field is absent the timeline frame is returned unchanged.
pub fn resolve_source_frame(range: &LayerRange, timeline_frame: usize) -> usize {
    match (range.source_in_frame, range.source_out_frame) {
        (Some(src_in), Some(src_out)) => {
            let offset = timeline_frame.saturating_sub(range.start_frame);
            (src_in + offset).min(src_out)
        }
        _ => timeline_frame,
    }
}

/// Returns the source frame index that should be passed to the video decoder
/// for the given timeline `frame_index`.
///
/// Iterates over all tracks and finds the first active range that has both
/// `source_in_frame` and `source_out_frame` set. If no such range exists the
/// original `frame_index` is returned unchanged.
pub fn resolve_decode_frame(tracks: &[LayerTrack], frame_index: usize) -> usize {
    for track in tracks {
        if let Some(range) = find_active_range(&track.ranges, frame_index) {
            if range.source_in_frame.is_some() && range.source_out_frame.is_some() {
                return resolve_source_frame(range, frame_index);
            }
        }
    }
    frame_index
}

pub fn apply_layer_tracks(
    layers: &[EffectLayer],
    tracks: &[LayerTrack],
    frame_index: usize,
) -> Vec<EffectLayer> {
    if tracks.is_empty() {
        return layers.to_vec();
    }

    layers.iter().map(|layer| {
        let mut modified_layer = layer.clone();
        
        // Find track for this layer
        if let Some(track) = tracks.iter().find(|t| t.layer_id == layer.id) {
            let active_range = find_active_range(&track.ranges, frame_index);
            let disable_outside = track.disable_outside_ranges.unwrap_or(true);

            if active_range.is_none() && disable_outside {
                modified_layer.enabled = false;
                return modified_layer;
            }

            // Apply range overrides
            if let Some(range) = active_range {
                if let Some(enabled) = range.enabled {
                    modified_layer.enabled = enabled;
                }
                if let Some(opacity) = range.opacity01 {
                    modified_layer.opacity = Some(opacity);
                }
                if let Some(intensity) = range.intensity {
                    modified_layer.intensity = intensity;
                }
            }

            // Apply keyframe interpolation
            let interpolated = interpolate_keyframes(&track.keyframes, frame_index);
            if let Some(opacity) = interpolated.opacity01 {
                modified_layer.opacity = Some(opacity);
            }
            if let Some(intensity) = interpolated.intensity {
                modified_layer.intensity = intensity;
            }
        }
        
        modified_layer
    }).collect()
}

fn find_active_range(ranges: &[LayerRange], frame: usize) -> Option<&LayerRange> {
    ranges.iter().find(|r| frame >= r.start_frame && frame <= r.end_frame)
}

struct InterpolatedValues {
    opacity01: Option<f32>,
    intensity: Option<f32>,
}

fn interpolate_keyframes(keyframes: &[LayerKeyframe], frame: usize) -> InterpolatedValues {
    if keyframes.is_empty() {
        return InterpolatedValues { opacity01: None, intensity: None };
    }

    let (prev, next) = find_keyframe_bounds(keyframes, frame);
    
    match (prev, next) {
        (None, None) => InterpolatedValues { opacity01: None, intensity: None },
        (Some(p), None) => InterpolatedValues {
            opacity01: p.opacity01,
            intensity: p.intensity,
        },
        (None, Some(n)) => InterpolatedValues {
            opacity01: n.opacity01,
            intensity: n.intensity,
        },
        (Some(p), Some(n)) => {
            if p.frame == n.frame {
                return InterpolatedValues {
                    opacity01: p.opacity01,
                    intensity: p.intensity,
                };
            }

            let span = (n.frame - p.frame) as f32;
            let t = (frame - p.frame) as f32 / span;
            
            let opacity01 = if let (Some(p_op), Some(n_op)) = (p.opacity01, n.opacity01) {
                Some(lerp(p_op, n_op, t).clamp(0.0, 1.0))
            } else {
                None
            };

            let intensity = if let (Some(p_int), Some(n_int)) = (p.intensity, n.intensity) {
                Some(lerp(p_int, n_int, t))
            } else {
                None
            };

            InterpolatedValues { opacity01, intensity }
        }
    }
}

fn find_keyframe_bounds(keyframes: &[LayerKeyframe], frame: usize) -> (Option<&LayerKeyframe>, Option<&LayerKeyframe>) {
    let mut prev: Option<&LayerKeyframe> = None;
    let mut next: Option<&LayerKeyframe> = None;

    for k in keyframes {
        if k.frame <= frame {
            if prev.is_none() || k.frame > prev.unwrap().frame {
                prev = Some(k);
            }
        }
        if k.frame >= frame {
            if next.is_none() || k.frame < next.unwrap().frame {
                next = Some(k);
            }
        }
    }

    (prev, next)
}

fn lerp(a: f32, b: f32, t: f32) -> f32 {
    a + (b - a) * t
}
